src/target/riscv/riscv-013.c

```diff
@@ -542,7 +553,7 @@ static dmi_status_t dmi_scan(struct target *target, uint32_t *address_in,
 
 	if (address_in)
 		*address_in = buf_get_u32(in, DTM_DMI_ADDRESS_OFFSET, info->abits);
-	dump_field(target, idle_count, &field);
+	dump_field(target, idle_count, &field, /*decode_in*/ data_in);
```

Comment:
I believe the condition for `decode_in` should be extended like this:

- We should keep track what type of DMI operation is currently in flight for the given debug module (e.g. store it to the instance `dm013_info_t`).
- The received data is only valid when when the current operation is DMI_OP_READ and at the same time the received `op==0 (success)`. 

In all other cases than the above, the received DMI data is irrelevant and should not be decoded. Decoding it would be misleading to the users who read the log.

Reply:
Currently, the condition is:
The value is decoded if it is not discarded by the caller of the function.

IMHO this is preferable, since it is more robust -- if there is an error in OpenOCD and a meaningless value is used and analyzed further in the code, such value will be present in the log. Moreover, the solution you suggest requires a lot of changes -- there is a separate interface to DMI operations (`batch.h/c`), which would need to update this field in `dm013_info_t`. If someone was to implement it -- the following still should be true: `!data_in || (<last op is read> && <in_op == success>)`. 
If you wish, I can add an assertion into `dump_field()`:
```
assert(!decode_in || in_op == 0)
```

Reply:
Thank you for replying. 

> If you wish, I can add an assertion into dump_field(): assert(!decode_in || in_op == 0)

Please don't. This assertion would not hold true when the answer from the target is `in_op==BUSY`.

> Currently, the condition is: The value is decoded if it is not discarded by the caller of the function. IMHO this is preferable, since it is more robust -- if there is an error in OpenOCD and a meaningless value is used and analyzed further in the code, such value will be present in the log.

Whether the received data is valid (and should be interpreted, for example decoded), is precisely determined by the JTAG DTM specification. That is, by the history of `op` values exchanged between OpenOCD and the target. Not by the fact that the caller of `dmi_scan()` reads the data_in.

In fact that the caller requests the data_in:
- does not mean the dmi.data is valid, 
- does not mean the caller will make any use of it.

> Moreover, the solution you suggest requires a lot of changes

I understand that it is more complex to implement it. It seems OK to me to skip this part for now, and, if you like, file a TODO issue for it.

For now, I believe you can at least partially improve the situation by not decoding the data when the target responds that the previous operation in unsuccessful:

```
	/* The write operation only got started if the previous operation succeeded. 
	 * Otherwise the write is ignored. */
	if (out_op == DTM_DMI_OP_WRITE && in_op == DMI_OP_SUCCESS) {
		/* ... */
	}
	/* The read data can only be valid if the previous operation succeeded. */	
	if (decode_in &&  in_op == DMI_OP_SUCCESS) {
		char in_decoded[decode_dmi(target, NULL, in_address, in_data) + 1];
		log_printf_lf(LOG_LVL_DEBUG, __FILE__, __LINE__, __PRETTY_FUNCTION__, "in: %s",
				decode_dmi(target, in_decoded, in_address, in_data)
				? in_decoded
				: "");
	}
```

Reply:
Addressed.

Reply:
Thanks.